formats/roam/roam.ts

```diff
@@ -0,0 +1,327 @@
+		let lineNumber = 0;
+
+		function processBlock(block: RoamBlock, level: number) {
+			lineNumber++;
+
+			if (block.uid) {
+				//check for roam DNP and convert to obsidian DNP
+				const dateObject = new Date(page.uid);
+				if (!isNaN(dateObject.getTime())) {
+					// The string can be converted to a Date object
+					const newPageTitle = convertDateString(page.title, userDNPFormat)
+					page.title = newPageTitle
+				}
+				// TODO why is lineNumber sometimes innaccurate?
+				const blockRefRegex = /.*?(\(\(.*?\)\)).*?/g
+				if (blockRefRegex.test(block.string)) {
+					toPostProcessblockLocations.set(block.uid, {
+						pageName: sanitizeFileNameKeepPath(page.title),
+						lineNumber: lineNumber,
+						blockString:block.string,
+					});
+				}
+				blockLocations.set(block.uid, {
+					pageName: sanitizeFileNameKeepPath(page.title),
+					lineNumber: lineNumber,
+					blockString:block.string,
+				});
+			}
+
+			if (block.children) {
+				for (let child of block.children) {
+				processBlock(child, level + 1);
+				}
+			}
+		}
+
+		if (page.children) {
+		for (let block of page.children) {
+			processBlock(block, 0);
+		}
+		}
+	}
+
+	return [blockLocations, toPostProcessblockLocations];
+}
+
+const roamMarkupScrubber = async (graphFolder:string, attachmentsFolder:string, blockText: string, downloadAttachments: boolean = false) => {
+	// get rid of roam-specific components
+	if (
+	  blockText.substring(0, 9) == "{{[[query" ||
+	  blockText.substring(0, 7) == "{{query"
+	)
+	  return "";
+	if (
+		blockText.substring(0, 10) == "{{[[streak" ||
+		blockText.substring(0, 8) == "{{streak"
+	)
+		return "";
+	if (blockText.substring(0, 12) == "{{attr-table") return "";
+	if (blockText.substring(0, 15) == "{{[[mentions]]:") return "";
+	if (blockText.substring(0, 8) == ":hiccup " && blockText.includes(":hr"))
+	  return "---"; // Horizontal line in markup, replace it with MD
+    if (blockText.substring(0, 9) == "{{[[search") return "";
+
+    //sanatize [[page names]]
+	//check for roam DNP and convert to obsidian DNP
+	blockText = blockText.replace(/\[\[(.*?)\]\]/g, (match, group1) => `[[${convertDateString(sanitizeFileNameKeepPath(group1), userDNPFormat)}]]`);
+	
+
+    // Regular expression to find nested pages [[SOME/TEXT]]     
+    // Replace each match with an Obsidian alias [[Artificial Intelligence|AI]]
+	blockText = blockText.replace(/\[\[(.*\/.*)\]\]/g, (_, group1) => `[[${graphFolder}/${group1}|${group1}]]`);
+	// regular block alias
+	blockText = blockText.replaceAll(/\[.+?\]\((\(.+?\)\))\)/g, "$1")
+	// page alias
+	blockText = blockText.replaceAll(/\[(.+?)\]\(\[\[(.+?)\]\]\)/g, "[[$2|$1]]")
+
+	blockText = blockText.replaceAll("[[>]]", ">");
+	blockText = blockText.replaceAll("{{TODO}}", "[ ]");
+	blockText = blockText.replaceAll("{{[[TODO]]}}", "[ ]");
+	blockText = blockText.replaceAll("{{DONE}}", "[x]");
+	blockText = blockText.replaceAll("{{[[DONE]]}}", "[x]");
+	blockText = blockText.replaceAll("{{word-count}}", "");
```

Comment:
Could we loop all of these that remove everything into a single regex so that we don't create a bunch of intermediate strings?

Reply:
Not sure what you're looking for here. Do you mean create a single regex to replace things like these

{{word-count}}
{{date}}
{{[[POMO]]}}
{{POMO}}
{{slider}}
{{[[slider]]}}
{{encrypt}}
{{[[encrypt]]}}
{{TaoOfRoam}}
{{orphans}}
{{[[orphans]]}}
{{count}}
{{character-count}}
{{comment-button}}

But avoid strings that have custom replacements like these?
{{TODO}}
{{[[TODO]]}}
{{DONE}}
{{[[DONE]]}}
{{embed: ((asdf))}}
{{[[embed]]: [[asadf]]}}

Reply:
Yeah the custom ones that are replaced with `[ ]` and `[x]` I guess you can keep, but I think it would be best to use a single regex for the other ones that just get replaced with empty string.